### PR TITLE
Draft of LC3 backend

### PR DESCRIPTION
**Pull request by ausbin**
lots of hacks in here but it does work on gradescope

thanks henry for your help and hard work here

### CODE REVIEW

File: zucchini/graders/pylc3_grader.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                'pip install parameterized']
+
+    def part_from_config_dict(self, config_dict):
+        return PyLC3Test.from_config_dict(config_dict)
+
+    def grade(self, submission, path, parts):
+        cmdline = ['python2', 'runner.py', self.test_class]
+        env = os.environ.copy()
+        # HACK HACK HACK HACK HACK
+        env['LD_LIBRARY_PATH'] = '/usr/local/lib/python2.7/dist-packages/pyLC3'
```
Comment (ausbin) [inline zucchini/graders/pylc3_grader.py]: This feels pretty bad. We need to not hardcode the path at least

File: zucchini/graders/pylc3_grader.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        if process.returncode != 0:
+            raise BrokenSubmissionError(
+                'grader command exited with nonzero exit code {}'
+                .format(process.returncode),
+                verbose=(process.stdout.decode()
+                         if process.stdout else '(no output)'))
+
+        with open(os.path.join(path, 'results.json'), 'r') as json_fp:
+            results = json.load(json_fp)
```
Comment (ausbin) [inline zucchini/graders/pylc3_grader.py]: need to throw a better error if the tester fails to produce the json

File: zucchini/graders/pylc3_grader.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    def grade(self, submission, path, parts):
+        cmdline = ['python2', 'runner.py', self.test_class]
+        env = os.environ.copy()
+        # HACK HACK HACK HACK HACK
+        env['LD_LIBRARY_PATH'] = '/usr/local/lib/python2.7/dist-packages/pyLC3'
+        try:
+            # Do not mix stderr into stdout because sometimes our friend
+            # Roi printStackTrace()s or System.err.println()s, and that
+            # will mess up JSON parsing
```
Comment (ausbin) [inline zucchini/graders/pylc3_grader.py]: outdated comment

**Comment by skyman** [comment]
I feel like we should probably merge this, especially if we're not planning to do upgrades on it, because I don't see us working on it after we're done with LC3 this sem haha

**Comment by dbecker1** [comment]
The environment hack has been removed as it is something we can configure from the Docker image. Will be building this branch and testing it on Docker before merging into master.

**Comment by skyman** [comment]
Let’s check if it works in a non-Docker environment too (e.g. in a fresh Ubuntu installation) since Gradescope is not the only use case for zucchini

**Comment by dbecker1** [comment]
Just tested it in a Docker image and a fresh install on Ubuntu 18.04 in Virtualbox. As long as the environment variable is set properly, it works fine. Not sure where/if we want to include it in the documentation since this is only specific to the pylc3 backend. 

Necessary variable:
LD_LIBRARY_PATH = '/usr/local/lib/python2.7/dist-packages/pyLC3'

**Comment by ausbin** [comment]
yeah it's my personal opinion that the right way forward is keeping environmental configuration like that ^ separate from the zucchini codebase. the 16.04 dependency packages hard-coded in the codebase kinda suck and i'm glad y'all are moving away from that

**Pull request merged**